File: articles/calculus-for-systems-modeling/articles/change-of-variables-in-multidimensional-systems/r/change_of_variables_audit.R
exposure_cartesian <- function(x, y) {
  r <- sqrt(x^2 + y^2)
  20 * exp(-0.4 * r)
}

exposure_polar <- function(r, theta) {
  20 * exp(-0.4 * r)
}

polar_total <- function(radius, radial_step, angular_step) {
  total <- 0
  r_values <- seq(radial_step / 2, radius, by = radial_step)
  theta_values <- seq(angular_step / 2, 2 * pi, by = angular_step)

  for (r in r_values) {
    for (theta in theta_values) {
      total <- total + exposure_polar(r, theta) * r * radial_step * angular_step
    }
  }

  total
}

cartesian_grid_total <- function(radius, step) {
  total <- 0
  xs <- seq(-radius, radius, by = step)
  ys <- seq(-radius, radius, by = step)

  for (x in xs) {
    for (y in ys) {
      if (x^2 + y^2 <= radius^2) {
        total <- total + exposure_cartesian(x, y) * step^2
      }
    }
  }

  total
}

audit_change_of_variables <- function(radius, radial_step, angular_step, scenario) {
  p_total <- polar_total(radius, radial_step, angular_step)
  c_total <- cartesian_grid_total(radius, radial_step)
  absolute_difference <- abs(p_total - c_total)
  relative_difference <- absolute_difference / max(abs(p_total), 1e-12)

  warning <- ifelse(
    radial_step > 0.5,
    "Resolution is coarse; transformed and Cartesian approximations may differ.",
    "Polar Jacobian factor r included; compare domain and resolution assumptions."
  )

  data.frame(
    scenario = scenario,
    radius = radius,
    radial_step = radial_step,
    angular_step = angular_step,
    polar_total = p_total,
    cartesian_grid_total = c_total,
    absolute_difference = absolute_difference,
    relative_difference = relative_difference,
    jacobian_rule = "dA = r dr dtheta",
    warning = warning
  )
}

results <- rbind(
  audit_change_of_variables(3.0, 0.5, pi / 24, "medium_polar_grid"),
  audit_change_of_variables(3.0, 0.25, pi / 48, "fine_polar_grid"),
  audit_change_of_variables(3.0, 0.125, pi / 96, "very_fine_polar_grid")
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_change_of_variables_audit.csv", row.names = FALSE)
print(results)

File: articles/calculus-for-systems-modeling/articles/change-of-variables-in-multidimensional-systems/julia/change_of_variables_audit.jl
exposure_cartesian(x, y) = begin
    r = sqrt(x*x + y*y)
    20.0 * exp(-0.4 * r)
end
exposure_polar(r, theta) = 20.0 * exp(-0.4 * r)

function polar_total(radius, radial_step, angular_step)
    total = 0.0
    r = radial_step / 2.0
    while r < radius
        theta = angular_step / 2.0
        while theta < 2.0*pi
            total += exposure_polar(r, theta) * r * radial_step * angular_step
            theta += angular_step
        end
        r += radial_step
    end
    return total
end

function cartesian_grid_total(radius, step)
    total = 0.0
    n = Int(floor((2.0 * radius) / step))
    for i in 0:n
        x = -radius + i * step
        for j in 0:n
            y = -radius + j * step
            if x*x + y*y <= radius*radius
                total += exposure_cartesian(x, y) * step * step
            end
        end
    end
    return total
end

function audit(radius, radial_step, angular_step, scenario)
    p_total = polar_total(radius, radial_step, angular_step)
    c_total = cartesian_grid_total(radius, radial_step)
    diff = abs(p_total - c_total)
    rel = diff / max(abs(p_total), 1.0e-12)
    warning = radial_step > 0.5 ? "Resolution is coarse; transformed and Cartesian approximations may differ." : "Polar Jacobian factor r included; compare domain and resolution assumptions."
    return scenario, radius, radial_step, angular_step, p_total, c_total, diff, rel, "dA = r dr dtheta", warning
end

println("scenario,radius,radial_step,angular_step,polar_total,cartesian_grid_total,absolute_difference,relative_difference,jacobian_rule,warning")
for case in [(3.0,0.5,pi/24,"medium_polar_grid"),(3.0,0.25,pi/48,"fine_polar_grid"),(3.0,0.125,pi/96,"very_fine_polar_grid")]
    println(join(audit(case[1], case[2], case[3], case[4]), ","))
end
